In [7]:
# Import required libraries
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import matplotlib.pyplot as plt
import hvplot.pandas

In [8]:
# Read in the clean spotify data
file_path = Path("Resources/spotify_clean.csv")
df_spotify = pd.read_csv(file_path)

# Display sample data
df_spotify.head()

track_name      track_artist  \
0  I Don't Care (with Justin Bieber) - Loud Luxur...        Ed Sheeran   
1                    Memories - Dillon Francis Remix          Maroon 5   
2                    All the Time - Don Diablo Remix      Zara Larsson   
3                  Call You Mine - Keanu Silva Remix  The Chainsmokers   
4            Someone You Loved - Future Humans Remix     Lewis Capaldi   

  track_album_release_date  danceability    energy       key  loudness  mode  \
0               2019-06-14      0.642049  1.201614  0.173200  1.367123     1   
1               2019-12-13      0.490412  0.643317  1.557627  0.585766     1   
2               2019-07-05      0.138889  1.284529 -1.211227  1.100090     0   
3               2019-07-19      0.435271  1.279002  0.450085  0.984309     1   
4               2019-03-05     -0.033426  0.742815 -1.211227  0.685151     1   

   speechiness  acousticness  ...  playlist_subgenre_new jack swing  \
0    -0.481362     -0.333898  ...                                 0   
1    -0.688642     -0.468670  ...                                 0   
2    -0.324422     -0.436799  ...                                 0   
3    -0.050024     -0.667642  ...                                 0   
4    -0.702460     -0.432701  ...                                 0   

   playlist_subgenre_permanent wave  playlist_subgenre_pop edm  \
0                                 0                          0   
1                                 0                          0   
2                                 0                          0   
3                                 0                          0   
4                                 0                          0   

   playlist_subgenre_post-teen pop  \
0                                0   
1                                0   
2                                0   
3                                0   
4                                0   

   playlist_subgenre_progressive electro house  playlist_subgenre_reggaeton  \
0                                            0                            0   
1                                            0                            0   
2                                            0                            0   
3                                            0                            0   
4                                            0                            0   

   playlist_subgenre_southern hip hop  playlist_subgenre_trap  \
0                                   0                       0   
1                                   0                       0   
2                                   0                       0   
3                                   0                       0   
4                                   0                       0   

   playlist_subgenre_tropical  playlist_subgenre_urban contemporary  
0                           0                                     0  
1                           0                                     0  
2                           0                                     0  
3                           0                                     0  
4                           0                                     0  

[5 rows x 47 columns]

In [9]:
df_spotify = df_spotify.set_index(['track_name', 'track_artist'])
df_spotify.head()

track_album_release_date  \
track_name                                         track_artist                                
I Don't Care (with Justin Bieber) - Loud Luxury... Ed Sheeran                     2019-06-14   
Memories - Dillon Francis Remix                    Maroon 5                       2019-12-13   
All the Time - Don Diablo Remix                    Zara Larsson                   2019-07-05   
Call You Mine - Keanu Silva Remix                  The Chainsmokers               2019-07-19   
Someone You Loved - Future Humans Remix            Lewis Capaldi                  2019-03-05   

                                                                     danceability  \
track_name                                         track_artist                     
I Don't Care (with Justin Bieber) - Loud Luxury... Ed Sheeran            0.642049   
Memories - Dillon Francis Remix                    Maroon 5              0.490412   
All the Time - Don Diablo Remix                    Zara Larsson          0.138889   
Call You Mine - Keanu Silva Remix                  The Chainsmokers      0.435271   
Someone You Loved - Future Humans Remix            Lewis Capaldi        -0.033426   

                                                                       energy  \
track_name                                         track_artist                 
I Don't Care (with Justin Bieber) - Loud Luxury... Ed Sheeran        1.201614   
Memories - Dillon Francis Remix                    Maroon 5          0.643317   
All the Time - Don Diablo Remix                    Zara Larsson      1.284529   
Call You Mine - Keanu Silva Remix                  The Chainsmokers  1.279002   
Someone You Loved - Future Humans Remix            Lewis Capaldi     0.742815   

                                                                          key  \
track_name                                         track_artist                 
I Don't Care (with Justin Bieber) - Loud Luxury... Ed Sheeran        0.173200   
Memories - Dillon Francis Remix                    Maroon 5          1.557627   
All the Time - Don Diablo Remix                    Zara Larsson     -1.211227   
Call You Mine - Keanu Silva Remix                  The Chainsmokers  0.450085   
Someone You Loved - Future Humans Remix            Lewis Capaldi    -1.211227   

                                                                     loudness  \
track_name                                         track_artist                 
I Don't Care (with Justin Bieber) - Loud Luxury... Ed Sheeran        1.367123   
Memories - Dillon Francis Remix                    Maroon 5          0.585766   
All the Time - Don Diablo Remix                    Zara Larsson      1.100090   
Call You Mine - Keanu Silva Remix                  The Chainsmokers  0.984309   
Someone You Loved - Future Humans Remix            Lewis Capaldi     0.685151   

                                                                     mode  \
track_name                                         track_artist             
I Don't Care (with Justin Bieber) - Loud Luxury... Ed Sheeran           1   
Memories - Dillon Francis Remix                    Maroon 5             1   
All the Time - Don Diablo Remix                    Zara Larsson         0   
Call You Mine - Keanu Silva Remix                  The Chainsmokers     1   
Someone You Loved - Future Humans Remix            Lewis Capaldi        1   

                                                                     speechiness  \
track_name                                         track_artist                    
I Don't Care (with Justin Bieber) - Loud Luxury... Ed Sheeran          -0.481362   
Memories - Dillon Francis Remix                    Maroon 5            -0.688642   
All the Time - Don Diablo Remix                    Zara Larsson        -0.324422   
Call You Mine - Keanu Silva Remix                  The Chainsmokers    -0.050024   
Someone You Loved - Future Humans Remix            

In [10]:
# Set the X and y variables
X = df_spotify.drop(columns=['track_album_release_date', 'track_popularity', 'bangers'])
y = df_spotify['track_popularity']

In [11]:
def model_mse(X, y):
    # Split the data into training and testing sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

    # Create a model with scikit-learn
    model = LinearRegression()

    # Fit the data into the model
    model.fit(X_train, y_train)

    # Create a prediction variable
    popularity_prediction = model.predict(X_test)

    # Evaluate the model using Mean Squared Error
    mse = mean_squared_error(y_test, popularity_prediction)
    return f'This model produces a Mean Squared Error of: {mse}'

model_mse(X,y)

'This model produces a Mean Squared Error of: 530.6364993985394'

In [12]:
df_spotify.columns

Index(['track_album_release_date', 'danceability', 'energy', 'key', 'loudness',
       'mode', 'speechiness', 'acousticness', 'instrumentalness', 'liveness',
       'valence', 'tempo', 'duration_ms', 'track_popularity', 'bangers',
       'playlist_genre_edm', 'playlist_genre_latin', 'playlist_genre_pop',
       'playlist_genre_r&b', 'playlist_genre_rap', 'playlist_genre_rock',
       'playlist_subgenre_album rock', 'playlist_subgenre_big room',
       'playlist_subgenre_classic rock', 'playlist_subgenre_dance pop',
       'playlist_subgenre_electro house', 'playlist_subgenre_electropop',
       'playlist_subgenre_gangster rap', 'playlist_subgenre_hard rock',
       'playlist_subgenre_hip hop', 'playlist_subgenre_hip pop',
       'playlist_subgenre_indie poptimism', 'playlist_subgenre_latin hip hop',
       'playlist_subgenre_latin pop', 'playlist_subgenre_neo soul',
       'playlist_subgenre_new jack swing', 'playlist_subgenre_permanent wave',
       'playlist_subgenre_pop edm', 'playl

In [13]:
music_variables = ['danceability', 'energy', 'key', 'loudness',
       'mode', 'speechiness', 'acousticness', 'instrumentalness', 'liveness',
       'valence', 'tempo', 'duration_ms']

model_mse(df_spotify[music_variables], df_spotify['track_popularity'])

'This model produces a Mean Squared Error of: 573.7720956487855'

In [14]:
all_genres = ['playlist_genre_edm', 'playlist_genre_latin', 'playlist_genre_pop',
       'playlist_genre_r&b', 'playlist_genre_rap', 'playlist_genre_rock',
       'playlist_subgenre_album rock', 'playlist_subgenre_big room',
       'playlist_subgenre_classic rock', 'playlist_subgenre_dance pop',
       'playlist_subgenre_electro house', 'playlist_subgenre_electropop',
       'playlist_subgenre_gangster rap', 'playlist_subgenre_hard rock',
       'playlist_subgenre_hip hop', 'playlist_subgenre_hip pop',
       'playlist_subgenre_indie poptimism', 'playlist_subgenre_latin hip hop',
       'playlist_subgenre_latin pop', 'playlist_subgenre_neo soul',
       'playlist_subgenre_new jack swing', 'playlist_subgenre_permanent wave',
       'playlist_subgenre_pop edm', 'playlist_subgenre_post-teen pop',
       'playlist_subgenre_progressive electro house',
       'playlist_subgenre_reggaeton', 'playlist_subgenre_southern hip hop',
       'playlist_subgenre_trap', 'playlist_subgenre_tropical',
       'playlist_subgenre_urban contemporary']

model_mse(df_spotify[all_genres], df_spotify['track_popularity'])

'This model produces a Mean Squared Error of: 550.5613283331905'

In [15]:
main_genres =['playlist_genre_edm', 'playlist_genre_latin', 'playlist_genre_pop',
       'playlist_genre_r&b', 'playlist_genre_rap', 'playlist_genre_rock']

model_mse(df_spotify[main_genres], df_spotify['track_popularity'])

'This model produces a Mean Squared Error of: 600.34934593957'

In [16]:
subgenres = ['playlist_subgenre_album rock', 'playlist_subgenre_big room',
       'playlist_subgenre_classic rock', 'playlist_subgenre_dance pop',
       'playlist_subgenre_electro house', 'playlist_subgenre_electropop',
       'playlist_subgenre_gangster rap', 'playlist_subgenre_hard rock',
       'playlist_subgenre_hip hop', 'playlist_subgenre_hip pop',
       'playlist_subgenre_indie poptimism', 'playlist_subgenre_latin hip hop',
       'playlist_subgenre_latin pop', 'playlist_subgenre_neo soul',
       'playlist_subgenre_new jack swing', 'playlist_subgenre_permanent wave',
       'playlist_subgenre_pop edm', 'playlist_subgenre_post-teen pop',
       'playlist_subgenre_progressive electro house',
       'playlist_subgenre_reggaeton', 'playlist_subgenre_southern hip hop',
       'playlist_subgenre_trap', 'playlist_subgenre_tropical',
       'playlist_subgenre_urban contemporary']

model_mse(df_spotify[subgenres], df_spotify['track_popularity'])

'This model produces a Mean Squared Error of: 550.3990780769535'

In [17]:
# Create a function that creates a LinearRegression model 
# and stores the model metrics in a DataFrame

def variable_test(X,y):
    # Reshape the X array
    X_reshaped = X.values.reshape(-1,1)

    # Create a model with scikit-learn
    model = LinearRegression()
    
    # Fit the data into the model
    model.fit(X_reshaped, y)

    # Make predictions using the X set
    predicted_y_values = model.predict(X_reshaped)

    # Compute metrics for the linear regression model: score, r2, mse, rmse, std
    r2 = r2_score(y, predicted_y_values)
    mse = mean_squared_error(y, predicted_y_values)
    rmse = np.sqrt(mse)
    std = np.std(y)

    # Store the metrics in a DataFrame
    model_variables = {}
    model_variables['r-squared'] = r2
    model_variables['Mean_Squared_Error'] = mse
    model_variables['Root_MSE'] = rmse
    model_variables['StandardDeviation'] = std

    metrics_df_list.append(model_variables)

    # Print relevant metrics.
    print(f"The r2 is {r2}.")
    print(f"The mean squared error is {mse}.")
    print(f"The root mean squared error is {rmse}.")
    print(f"The standard deviation is {std}.")

In [28]:
# Create a list of just the music variables
music_variables = ['danceability', 'energy', 'key', 'loudness',
       'mode', 'speechiness', 'acousticness', 'instrumentalness', 'liveness',
       'valence', 'tempo', 'duration_ms']

# Define an empty list to store the dictionaries produced by the 
# variable_test function

metrics_df_list = []

for var in music_variables:
    print(f"\n----------{var}----------")
    
    result = variable_test(df_spotify[var], df_spotify['track_popularity'])
    print(result)
    
    song_plot = df_spotify.hvplot.scatter(
       x= 'track_popularity',
       y= var,
       title= f"{var.title()} vs. Track Popularity"
       )
    song_plot


----------danceability----------
The r2 is 0.004192260933750247.
The mean squared error is 621.5682201993782.
The root mean squared error is 24.93126992753033.
The standard deviation is 24.98369401128743.
None

----------energy----------
The r2 is 0.011905326535265326.
The mean squared error is 616.7538406056724.
The root mean squared error is 24.834529200403065.
The standard deviation is 24.98369401128743.
None

----------key----------
The r2 is 4.229594385840585e-07.
The mean squared error is 624.1847024447362.
The root mean squared error is 24.98368872774267.
The standard deviation is 24.98369401128743.
None

----------loudness----------
The r2 is 0.003327798902697743.
The mean squared error is 622.1078044032276.
The root mean squared error is 24.942089014419533.
The standard deviation is 24.98369401128743.
None

----------mode----------
The r2 is 0.00011313675308255444.
The mean squared error is 624.1143481892321.
The root mean squared error is 24.982280684301667.
The standard dev

In [31]:
for var in music_variables:
    song_plot = df_spotify.hvplot.scatter(
       x= 'track_popularity',
       y= var,
       title= f"{var.title()} vs. Track Popularity"
       )
    song_plot

In [ ]:
# Using the list of dictionaries, create a DateFrame
metrics_df = pd.DataFrame(metrics_df_list)
metrics_df

r-squared  Mean_Squared_Error   Root_MSE  StandardDeviation
0   4.192261e-03          621.568220  24.931270          24.983694
1   1.190533e-02          616.753841  24.834529          24.983694
2   4.229594e-07          624.184702  24.983689          24.983694
3   3.327799e-03          622.107804  24.942089          24.983694
4   1.131368e-04          624.114348  24.982281          24.983694
5   4.650479e-05          624.155939  24.983113          24.983694
6   7.252113e-03          619.658307  24.892937          24.983694
7   2.246174e-02          610.164686  24.701512          24.983694
8   2.979461e-03          622.325232  24.946447          24.983694
9   1.104321e-03          623.495666  24.969895          24.983694
10  2.892356e-05          624.166913  24.983333          24.983694
11  2.064462e-02          611.298907  24.724460          24.983694

In [ ]:
# Add the music varialbe to the DateFrame
metrics_df['metric'] = music_variables

# Re-order the DataFrame
metrics_df= metrics_df[['metric', 'r-squared', 'Mean_Squared_Error', 'Root_MSE', 'StandardDeviation']]
metrics_df

metric     r-squared  Mean_Squared_Error   Root_MSE  \
0       danceability  4.192261e-03          621.568220  24.931270   
1             energy  1.190533e-02          616.753841  24.834529   
2                key  4.229594e-07          624.184702  24.983689   
3           loudness  3.327799e-03          622.107804  24.942089   
4               mode  1.131368e-04          624.114348  24.982281   
5        speechiness  4.650479e-05          624.155939  24.983113   
6       acousticness  7.252113e-03          619.658307  24.892937   
7   instrumentalness  2.246174e-02          610.164686  24.701512   
8           liveness  2.979461e-03          622.325232  24.946447   
9            valence  1.104321e-03          623.495666  24.969895   
10             tempo  2.892356e-05          624.166913  24.983333   
11       duration_ms  2.064462e-02          611.298907  24.724460   

    StandardDeviation  
0           24.983694  
1           24.983694  
2           24.983694  
3           24.983694  
4           24.983694  
5           24.983694  
6           24.983694  
7           24.983694  
8           24.983694  
9           24.983694  
10          24.983694  
11          24.983694

# Analysis
Models were run using: 
- All music variables, all genres, and sub-genres (MSE: 530)
- Just music variables (MSE: 574)
- All genres/sub-genres (MSE: 551)
- Just genres (MSE: 600)
- Just sub-genres (MSE: 550)
- Each music variable (MSE: ranged 610-624)

The best model was the model including all variables. However, the model does not do a good job of predicting the popularity of a song. The model, on average, comes within 23 points of the actual popularity score of a song. The scale is 0-100 which means that on average, the score is off by a quarter the posible range.

A Linear Regression model is not a good fit for predicting a song's popularity.